File: backend/app/api/seed.py
"""Planting the seed.

Two steps, because the screen has two (§12, §48): the layers are supplied
one at a time and each is parsed the moment it lands, then initialization
commits all three at once.

Both endpoints take the file *text*. The browser reads the dropped file
and posts its contents, which keeps the seam free of multipart handling
and, more usefully, keeps parsing stateless: nothing is held between the
parse of a layer and the initialization that commits it. The frontend
holds the three files until the operator initializes, and System State
remains the only server-side state there is (FR-L5).
"""

from fastapi import APIRouter

from app.domain.events import Category
from app.domain.lifecycle import LifecycleState, assert_transition
from app.domain.schema import Schema
from app.knowledge.seed_loader import (
    LAYER_FILENAMES,
    LAYER_ORDER,
    LAYER_ROLES,
    Layer,
    LayerSummary,
    load_layer,
    load_seed,
    read_bundled,
)
from app.domain.state import StateSnapshot
from app.runtime import state

router = APIRouter(prefix="/seed")


class LayerUpload(Schema):
    """One supplied layer, as text."""

    layer: Layer
    content: str


class SeedUpload(Schema):
    """All three layers, supplied together at initialization (FR-S3)."""

    layers: dict[Layer, str]


class LayerDescriptor(Schema):
    """What the seed screen needs to draw one slot before anything is dropped."""

    layer: Layer
    filename: str
    role: str


@router.get("/layers", response_model=list[LayerDescriptor])
async def get_layers() -> list[LayerDescriptor]:
    """The three slots, in order, with the question each layer answers."""
    return [
        LayerDescriptor(layer=layer, filename=LAYER_FILENAMES[layer], role=LAYER_ROLES[layer])
        for layer in LAYER_ORDER
    ]


@router.post("/parse", response_model=LayerSummary)
async def parse(upload: LayerUpload) -> LayerSummary:
    """Validate and summarise one layer, without committing to it.

    This is what makes the per-layer summary on the seed screen real
    (FR-S5): the headings reported are the headings in the file that was
    just dropped. A file that cannot be read as Markdown is refused here,
    with the reason (FR-S4).
    """
    return load_layer(upload.layer, upload.content)


@router.get("/bundled", response_model=SeedUpload)
async def get_bundled() -> SeedUpload:
    """The seed files shipped with the repository (FR-S7).

    The hidden affordance fetches these and supplies them exactly as a
    dropped file would be, so the shortcut skips the file picker and
    nothing else.
    """
    return SeedUpload(layers=read_bundled())


@router.post("/initialize", response_model=StateSnapshot)
async def initialize(upload: SeedUpload) -> StateSnapshot:
    """Plant the seed and initialize the system (FR-S2, FR-S3).

    All three layers are validated before anything is recorded, so a
    rejected seed leaves the system exactly as it was. The transition is
    the last thing to happen, and it is what moves the interface off the
    seed screen.
    """
    # Checked before anything is parsed or recorded, so a second plant is
    # refused by the state machine without leaving a seed behind (FR-L3).
    assert_transition(state.lifecycle, LifecycleState.INITIALIZED)

    summary = load_seed(upload.layers)
    state.seed = summary.model_dump(by_alias=True)

    state.record(
        type="seed.loaded",
        category=Category.SUCCESS,
        message=(
            f"Seed accepted. Three layers registered from {summary.heading_count} "
            "parsed headings."
        ),
        payload={
            "layers": [
                {
                    "layer": entry.layer,
                    "filename": entry.filename,
                    "title": entry.title,
                    "headings": entry.heading_count,
                    "sections": entry.sections,
                }
                for entry in summary.layers
            ]
        },
    )

    state.transition(LifecycleState.INITIALIZED)
    return state.snapshot()

File: frontend/src/stores/seed.ts
import { computed, ref } from 'vue'
import { defineStore } from 'pinia'

export type Layer = 'core' | 'adaptation' | 'protection'

export interface Heading {
  level: number
  text: string
}

export interface LayerSummary {
  layer: Layer
  filename: string
  title: string | null
  headings: Heading[]
  headingCount: number
  sections: number
  topics: number
  lines: number
  characters: number
}

export interface LayerDescriptor {
  layer: Layer
  filename: string
  role: string
}

/** One of the three slots on the seed screen, and what has landed in it. */
export interface Slot {
  descriptor: LayerDescriptor
  /** The name of the file the operator supplied, which need not be the canonical one. */
  supplied: string | null
  summary: LayerSummary | null
  error: string | null
  busy: boolean
}

interface Rejection {
  detail: string
  layer?: Layer
  reason?: string
}

/**
 * The plant (FR-S2).
 *
 * Two steps, because the gesture has two. A layer is parsed the moment it
 * lands, so the summary beside it is derived from the file that was
 * actually supplied rather than from a fixture (FR-S5). Initialization
 * then commits all three at once, which is where the lifecycle moves.
 *
 * The file text is held here until then. It is never parsed in the
 * browser: a summary the frontend invented would look identical and mean
 * nothing.
 */
export const useSeedStore = defineStore('seed', () => {
  const slots = ref<Slot[]>([])
  const initializing = ref(false)
  const failure = ref<string | null>(null)

  /** Supplied file text, keyed by layer, held until initialization. */
  const staged = new Map<Layer, string>()

  const ready = computed(
    () => slots.value.length === 3 && slots.value.every((slot) => slot.summary !== null),
  )

  const count = computed(() => slots.value.filter((slot) => slot.summary !== null).length)

  function slotOf(layer: Layer): Slot | undefined {
    return slots.value.find((slot) => slot.descriptor.layer === layer)
  }

  async function fetchLayers(): Promise<void> {
    const response = await fetch('/api/seed/layers')
    if (!response.ok) {
      failure.value = `The seed layers could not be read: ${response.status}`
      return
    }
    const descriptors = (await response.json()) as LayerDescriptor[]
    slots.value = descriptors.map((descriptor) => ({
      descriptor,
      supplied: null,
      summary: null,
      error: null,
      busy: false,
    }))
  }

  /**
   * Supply one layer and parse it (FR-S4, FR-S5).
   *
   * A rejection clears the slot rather than leaving a stale summary
   * beside a file that was refused.
   */
  async function supply(layer: Layer, filename: string, content: string): Promise<void> {
    const slot = slotOf(layer)
    if (!slot) {
      return
    }

    slot.busy = true
    slot.error = null
    try {
      const response = await fetch('/api/seed/parse', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ layer, content }),
      })

      if (!response.ok) {
        const rejection = (await response.json()) as Rejection
        slot.summary = null
        slot.supplied = filename
        slot.error = rejection.reason ?? rejection.detail
        staged.delete(layer)
        return
      }

      slot.summary = (await response.json()) as LayerSummary
      slot.supplied = filename
      staged.set(layer, content)
    } finally {
      slot.busy = false
    }
  }

  /** Read a dropped or chosen file and supply it. */
  async function supplyFile(layer: Layer, file: File): Promise<void> {
    await supply(layer, file.name, await file.text())
  }

  function clear(layer: Layer): void {
    const slot = slotOf(layer)
    if (slot) {
      slot.supplied = null
      slot.summary = null
      slot.error = null
    }
    staged.delete(layer)
  }

  /**
   * Load the seed files shipped with the repository (FR-S7).
   *
   * The hidden affordance skips the file picker and nothing else: the
   * text returned here goes through the same parse as a dropped file, so
   * the screen fills in exactly as it would have.
   */
  async function loadBundled(): Promise<void> {
    failure.value = null
    const response = await fetch('/api/seed/bundled')
    if (!response.ok) {
      failure.value = `The bundled seed could not be read: ${response.status}`
      return
    }

    const { layers } = (await response.json()) as { layers: Record<Layer, string> }
    for (const slot of slots.value) {
      const layer = slot.descriptor.layer
      await supply(layer, slot.descriptor.filename, layers[layer])
    }
  }

  /**
   * Plant the seed (FR-S3).
   *
   * The lifecycle change is not adopted from the response. It arrives on
   * the event stream like every other transition, so the screen that
   * replaces this one is driven by the same events the audit log is
   * built from (FR-E5).
   */
  async function initialize(): Promise<void> {
    if (!ready.value || initializing.value) {
      return
    }

    initializing.value = true
    failure.value = null
    try {
      const response = await fetch('/api/seed/initialize', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ layers: Object.fromEntries(staged) }),
      })

      if (!response.ok) {
        const rejection = (await response.json()) as Rejection
        failure.value = rejection.detail
      }
    } finally {
      initializing.value = false
    }
  }

  /** Forget the staged seed, so Reset returns to an empty screen (FR-O4). */
  function discard(): void {
    for (const slot of slots.value) {
      clear(slot.descriptor.layer)
    }
    failure.value = null
  }

  return {
    slots,
    initializing,
    failure,
    ready,
    count,
    fetchLayers,
    supply,
    supplyFile,
    clear,
    loadBundled,
    initialize,
    discard,
  }
})
